boards/beagle/beagleconnect_freedom/board_antenna.c

```diff
@@ -1,116 +1,119 @@
-/* SPDX-License-Identifier: Apache-2.0
- *
+/*
  * Copyright (c) 2021 Florin Stancu
- * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2024 Ayush Singh <ayush@beagleboard.org>
  *
+ * SPDX-License-Identifier: Apache-2.0
  */
 
 /*
  * Implements the RF driver callback to configure the on-board antenna
  * switch.
  */
 
-#include <zephyr/kernel.h>
-#include <zephyr/device.h>
+#define DT_DRV_COMPAT skyworks_sky13317
+
 #include <zephyr/init.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/pinctrl.h>
 
 #include <ti/drivers/rf/RF.h>
-#include <driverlib/gpio.h>
-#include <driverlib/ioc.h>
 #include <driverlib/rom.h>
+#include <driverlib/interrupt.h>
 
-/* DIOs for RF antenna paths */
-#define BOARD_RF_HIGH_PA   29		/* TODO: pull from DT */
-#define BOARD_RF_SUB1GHZ   30		/* TODO: pull from DT */
+/* custom pinctrl states for the antenna mux */
+#define PINCTRL_STATE_ANT_SUBG    1
+#define PINCTRL_STATE_ANT_SUBG_PA 2
 
-static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events,
-		void *arg);
+#define BOARD_ANT_GPIO_PA   0
+#define BOARD_ANT_GPIO_SUBG 1
 
+static int board_antenna_init(const struct device *dev);
+static void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg);
 
 const RFCC26XX_HWAttrsV2 RFCC26XX_hwAttrs = {
-	.hwiPriority        = INT_PRI_LEVEL7,
-	.swiPriority        = 0,
+	.hwiPriority = INT_PRI_LEVEL7,
+	.swiPriority = 0,
 	.xoscHfAlwaysNeeded = true,
 	/* RF driver callback for custom antenna switching */
 	.globalCallback = board_cc13xx_rf_callback,
 	/* Subscribe to events */
-	.globalEventMask = (RF_GlobalEventRadioSetup |
-			RF_GlobalEventRadioPowerDown),
+	.globalEventMask = (RF_GlobalEventRadioSetup | RF_GlobalEventRadioPowerDown),
 };
 
+PINCTRL_DT_INST_DEFINE(0);
+DEVICE_DT_INST_DEFINE(0, board_antenna_init, NULL, NULL, NULL, POST_KERNEL,
+		      CONFIG_BOARD_ANTENNA_INIT_PRIO, NULL);
```

Comment:
I like the approach of moving to DT initialization rather than kernel initialization to make it more generic.